client/blb/client.go

```diff
@@ -1118,6 +1118,104 @@ type tractResultRepl struct {
 	thisB []byte
 }
 
+func (cli *Client) readOneTractWithResult(
+	ctx context.Context,
+	host string,
+	reqID string,
+	tract *core.TractInfo,
+	thisB []byte,
+	thisOffset int64) tractResultRepl {
+	var badVersionHost string
+	// TODO(eric): fill in otherHosts when ts-ts cancellation is done.
+	thisBLoc, err := cli.tractservers.Read(ctx, host, nil, reqID, tract.Tract, tract.Version, len(thisB), thisOffset)
+	if err == core.ErrVersionMismatch {
+		// TODO(eric): fix this check when we redo error reporting.
+		badVersionHost = host
+	}
+	readDone()
+	if err != core.NoError && err != core.ErrEOF {
+		// TODO(eric): redo bad TS reporting mechanism.
+		return tractResultRepl{
+			tractResult: tractResult{0, 0, err, badVersionHost},
+		}
+	}
+	log.V(1).Infof("read %s from tractserver at address %s: %s", tract.Tract, host, err)
+
+	return tractResultRepl{
+		tractResult: tractResult{len(thisB), len(thisBLoc), err, badVersionHost},
+		thisB:       thisBLoc,
+	}
+}
+
+// vars for test injection
+var (
+	randOrder         = getRandomOrder
+	backupRequestFunc = doParallelBackupReads
+	readDone          = func() {} // call when read/readAt rpcs return.
+	backupPhaseDone   = func() {} // call when the entire backup read phase is done.
+)
+
+func (cli *Client) readOneTractBackupReplicated(
+	ctx context.Context,
+	reqID string,
+	tract *core.TractInfo,
+	thisB []byte,
+	thisOffset int64,
+	order []int,
+	n int,
+	resultCh chan tractResultRepl) {
+	err := core.ErrAllocHost // default error if none present
+	var badVersionHost string
+	host := tract.Hosts[order[n]]
+	if host == "" {
+		log.V(1).Infof("read %s from tsid %d: no host", tract.Tract, tract.TSIDs[n])
+		resultCh <- tractResultRepl{
+			tractResult: tractResult{0, 0, err, badVersionHost},
+		}
+		return
+	}
+	delay := cli.backupReadState.BackupReadBehavior.Delay
+	select {
+	case <-ctx.Done():
+		log.V(1).Infof("backup read %s-%d to tractserver cancelled", tract.Tract, n)
+		resultCh <- tractResultRepl{
+			tractResult: tractResult{0, 0, core.ErrCanceled, badVersionHost},
+		}
+	case <-cli.backupReadState.delayFunc(time.Duration(n) * delay):
```

Comment:
could pass in the delay value instead of n? or possibly even the timer channel itself? then you might not need delayFunc at all, since you're going to override the spawning logic.

Reply:
Are you thinking that we override spawning logic at the layer below by putting `readOneTractWithResult` calls in different goroutines? It would bypass the delayFunc usage in the select loop. I think bch was used to make sure firing of the timer could be controlled without using time.Sleep.

Reply:
I was thinking the latter.. pass in 0 or time.After(0) for all the calls in your backupRequestFunc implementation and control things by the order that you spawn them.

Reply:
Passing the delay channel seemed to work better for testing. Went with that over time.After(0).